File: frontend/oei-web/src/app/domain/model/wallet/wallet-pass.ts
export type WalletPassProvider = 'APPLE' | 'GOOGLE';

export type WalletPassStatus = 'MOCKED' | 'ISSUED' | 'REVOKED' | 'RENEWED';

// `mocked` is always `true` in V1 (ADR 0002 / OpenAPI): this represents the pass
// structure without generating a real signed `.pkpass` (no Apple/Google publisher
// certificate available yet). MUST NEVER be presented as an official ID document —
// enforced in the presentation layer with an explicit "démonstration" notice.
export interface WalletPass {
  readonly id: string;
  readonly memberId: string;
  readonly provider: WalletPassProvider;
  readonly status: WalletPassStatus;
  readonly serialNumber: string;
  readonly verificationUrl?: string;
  readonly levelColor?: string;
  readonly issuedAt: string;
  readonly revokedAt?: string | null;
  readonly mocked: true;
}

export function createWalletPass(fields: Omit<WalletPass, 'mocked'>): WalletPass {
  return Object.freeze({ ...fields, mocked: true });
}

export interface WalletPassVerification {
  readonly valid: boolean;
  readonly memberPublicSlug?: string;
  readonly status: WalletPassStatus;
  readonly tier?: string;
}

File: backend/domain/shared/src/main/java/global/oei/domain/shared/wallet/WalletPass.java
package global.oei.domain.shared.wallet;

import java.time.Instant;
import java.util.Objects;

import global.oei.domain.shared.member.MemberId;

/**
 * A member's Apple/Google Wallet pass. {@link #mocked()} is always {@code true} in this
 * iteration: no real signed {@code .pkpass} is generated (no publisher certificate
 * available) — see {@code CreateWalletPassService}, the sole place allowed to construct
 * one, which enforces this invariant. Never to be presented as an official identity
 * document.
 */
public record WalletPass(
        String id,
        MemberId memberId,
        WalletPassProvider provider,
        WalletPassStatus status,
        String serialNumber,
        String verificationUrl,
        String levelColor,
        Instant issuedAt,
        Instant revokedAt,
        boolean mocked) {

    public WalletPass {
        Objects.requireNonNull(id, "id must not be null");
        Objects.requireNonNull(memberId, "memberId must not be null");
        Objects.requireNonNull(provider, "provider must not be null");
        Objects.requireNonNull(status, "status must not be null");
        Objects.requireNonNull(serialNumber, "serialNumber must not be null");
    }

    /**
     * @return a new instance with {@link #status()} set to {@link WalletPassStatus#REVOKED}
     *         and {@link #revokedAt()} set to {@code revokedAt}; every other field unchanged
     */
    public WalletPass revoke(final Instant revokedAt) {
        return new WalletPass(
                id, memberId, provider, WalletPassStatus.REVOKED, serialNumber, verificationUrl, levelColor, issuedAt,
                revokedAt, mocked);
    }
}
